### PR TITLE
feat: group feature implementation

### PR DESCRIPTION
**Pull request by maze-runnar**
Related to https://github.com/fossasia/open-event-frontend/issues/5041

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by maze-runnar** [comment]
Screenshot from 2021-01-20 11-23-31

File: app/controllers/groups/create.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import Controller from '@ember/controller';
+import { action } from '@ember/object';
+import { inject as service } from '@ember/service';
+
+export default class extends Controller {
+
+  @service errorHandler;
+
+  GroupName ='';
```
Comment (iamareebjamal) [inline app/controllers/groups/create.js:9]: Sundaram, have you ever seen a JS variable, at least in this project starting with a capital letter?

File: app/controllers/groups/create.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      });
+      await myGroup.save();
+      this.notify.success(this.l10n.t('Your group has been saved'),
+        {
+          id: 'group_save'
+        });
+      this.router.transitionTo('groups.list');
+    } catch (e) {
+      console.error('Error while saving group', e);
+      const message = e.errors?.[0]?.detail ?? this.l10n.t('Oops something went wrong. Please try again');
```
Comment (iamareebjamal) [inline app/controllers/groups/create.js:27]: You created error handler service if I remember correctly, right?

**Comment by iamareebjamal** [comment]
Check out My Tickets and My Events section and tell me is it similar?

This is not admin section, this is user section feature

File: app/controllers/groups/create.js (hunk 1 above)
Comment (maze-runnar) [reply app/controllers/groups/create.js:9]: never 🤦🏻‍♂️

**Comment by maze-runnar** [comment]
> Check out `My Tickets` and `Manage Events` section and tell me is it similar?
> 
> This is not admin section, this is user section feature

Similar in what terms like code or functionality?   Some feature in user section like access-code, videoroom, discount-code are like this.

**Comment by maze-runnar** [comment]
instead of tab I should use a button for group creation?

**Comment by iamareebjamal** [comment]
See the UI

**Comment by iamareebjamal** [comment]
They are not tables

**Comment by maze-runnar** [comment]
Screenshot from 2021-01-22 18-03-05
Screenshot from 2021-01-22 17-58-16
Screenshot from 2021-01-22 17-58-09

File: app/controllers/groups/create.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import { action } from '@ember/object';
+import { inject as service } from '@ember/service';
+
+export default class extends Controller {
+
+  @service errorHandler;
+
+  groupName ='';
+
+  groupEvents = [];
```
Comment (iamareebjamal) [inline app/controllers/groups/create.js:11]: Why storing it here? You already have group model, Use `group.events`

That's why it is not switching correctly

File: app/controllers/groups/create.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import Controller from '@ember/controller';
+import { action } from '@ember/object';
+import { inject as service } from '@ember/service';
+
+export default class extends Controller {
+
+  @service errorHandler;
+
+  groupName ='';
```
Comment (iamareebjamal) [inline app/controllers/groups/create.js:9]: Use `group.name`

File: app/controllers/groups/create.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+  @action
+  async submit() {
+    try {
+      this.loading = true;
+      const myGroup = this.store.createRecord('group', {
+        user   : this.authManager.currentUser,
+        name   : this.groupName,
+        events : this.groupEvents
+      });
```
Comment (iamareebjamal) [inline app/controllers/groups/create.js:31]: Create the group in route itself. See how it is done in video stream route for create and edit

File: app/controllers/groups/create.js (hunk 4 above)
Comment (maze-runnar) [reply app/controllers/groups/create.js:9]: but i am creating it in controller, so on first time there is no model.group.name?

File: app/controllers/groups/create.js (hunk 5 above)
Comment (maze-runnar) [reply app/controllers/groups/create.js:31]: ok

Comment (maze-runnar) [reply app/controllers/groups/create.js:31]: is it wrong to create it in controller?

Comment (iamareebjamal) [reply app/controllers/groups/create.js:31]: Then you have to do things this way and it won't automatically do things we discussed in call

**Comment by maze-runnar** [comment]
Screenshot from 2021-01-24 22-15-13

File: app/controllers/groups/create.js [hunk 6]
```diff
@@ -6,15 +6,16 @@ export default class extends Controller {
...
-  groupName ='';
-
-  groupEvents = [];
-
   @action
   addEvent(event) {
-    if (!this.groupEvents.includes(event)) {
-      this.groupEvents.push(event);
-    }
+    this.model.group.set('events', [event].concat(this.model.group.events.toArray()));
```
Comment (iamareebjamal) [inline app/controllers/groups/create.js:11]: Use pushObject

Comment (maze-runnar) [reply app/controllers/groups/create.js:11]: ok

File: app/controllers/groups/create.js [hunk 7]
```diff
@@ -6,15 +6,16 @@ export default class extends Controller {
...
+  }
+
+  @action
+  removeEvent(event) {
+    let groupEvents = this.model.group.events.toArray();
+    groupEvents = groupEvents.filter(x => x !== event);
+    this.model.group.set('events', groupEvents);
```
Comment (iamareebjamal) [inline app/controllers/groups/create.js:18]: Use removeObject

**Comment by iamareebjamal** [comment]
As edit and create controller have same code, move the logic in GroupForm like in Video Stream route/controller

**Comment by maze-runnar** [comment]
> As edit and create controller have same code, move the logic in GroupForm like in Video Stream route/controller

ok

**Comment by iamareebjamal** [comment]
Cool, last thing would be to use glimmer component instead of classic

**Comment by maze-runnar** [comment]
> Cool, last thing would be to use glimmer component instead of classic

I tried that it is giving some error like `you are passing undefined in action without quotes`

**Comment by mariobehling** [comment]
Thanks. Please make the following changes

* On Event Pages, that are part of a group add a follow button below the calendar.
Screenshot from 2021-01-25 23-20-50

* On the Group Overview Page
  * Make a blue "Create Group" Button to the right of the screen (similar to other Create/Add buttons)
  * Change the following
     * Heading My Groups -> "Groups"
     * Tab Groups -> "My Groups" and show all groups of the account here
     * Tab Create -> "Following" and show all groups that the user follows
  * Move "Save" button above "All events"
  * The Group image has the default image "Event". Please change this to a default image "Group"

**Comment by mariobehling** [comment]
Please fix this:
On the "Create Group" or "Edit Group" page, when a user adds an event to the group from the section "All Events" the event gets added to the area "Group Events". Expected is: 
* That at the same time the event should not show up anymore in the section "All Events". 
* When a user removes and event from the group the event should then show up in the list under "All events" again though.



Screenshot from 2021-01-25 23-30-04

**Comment by iamareebjamal** [comment]
Follow button cannot be added right now because it is not implemented on the server

**Comment by iamareebjamal** [comment]
Since we don't have the source of the events placeholder image, we cannot change it to groups, so I have shown the landing image as banner

**Comment by iamareebjamal** [comment]
Other changes will be followed up in a PR after following feature is implemented

**Pull request merged**